docs/docs/recipes/20-gmd.md

```diff
@@ -0,0 +1,104 @@
+import OpenNew from '@site/static/img/open_new.svg';
+
+# Configuring Testify to run on Gradle managed device
+
+Starting from API levels
+27 [Android Gradle Plugin allows you to configure virtual test devices in your project's Gradle files <OpenNew />](https://developer.android.com/studio/test/gradle-managed-devices).
+This feature improves testing experience by delegating the task of starting, shutting down and
+managing emulator to the Gradle.
+
+When using Gradle managed device user must
+use [Test storage service <OpenNew />](https://developer.android.com/reference/androidx/test/services/storage/TestStorage)
+to save screenshots. Using this service Android Gradle Plugin would be able to save test output
+including screenshots and diffs to the `build` folder
+
+### Adding Gradle managed device configuration to build file
+
+To create new Gradle managed device add it's definition to the `build.gradle` file:
+
+```groovy
+
+android {
+    testOptions {
+        managedDevices {
+            devices {
+                NAME_OF_THE_DEVICE(com.android.build.api.dsl.ManagedVirtualDevice) {
+                    device = "Pixel 2"
+                    apiLevel = 30
+                    systemImageSource = "aosp"
+                }
+            }
+        }
+    }
+}
+
+```
+
+In order to receive screenshots after test execution enable `TestStorageService`:
+
+```groovy
+
+android {
+    defaultConfig {
+        testInstrumentationRunnerArguments useTestStorageService: "true"
+    }
+}
+
+dependencies {
+    androidTestUtil("androidx.test.services:test-services:1.4.2")
+}
+```
+
+With that changes applied couple of new Gradle tasks should be added to your project. Suppose, you
+named your device `tester`, then `testerCheck` and `testDebugAndroidTest` tasks should be added to
+your project.
+
+### Configuring the Testify library
+
+You also need to instruct Testify to use Test storage to store screenshots and diffs. This could be
+done using Testify plugin configuration:
+
+```groovy
+testify {
+    useTestStorage = true
+}
+```
+
+### Running screenshot tests
+
+To perform screenshot test verification run any of the new Gradle tasks, for
+example `./gradlew testerDebugAndroidTest`.
+
+Because we need to run specific Gradle task to execute our tests using Gradle managed device, we
+cannot use `screenshotPull` task. When execution completed go to the
+module's `build/outputs/managed_device_android_test_additional_output/tester` folder to get recorded
+screenshots or diffs.
+
+### Updating baselines
+
+Because we need to run specific Gradle task to execute our tests using Gradle managed device, we
+cannot use `screenshotRecord` task. In order to generate new baseline apply necessary setting to
+the `ScreenshotRule`:
+
+```kotlin
+@get:Rule
+var rule = ScreenshotRule(ClientListActivity::class.java)
+
+@ScreenshotInstrumentation
+@Test
+fun testMissingBaseline() {
+    rule
+        .setRecordModeEnabled(true)
```

Comment:
I think it would be preferable to set the record mode via a build.gradle flag, similar to `useTestStorage`.

e.g.

```
testify {
    isRecordMode System.getenv('TESTIFY_RECORD_MODE')
}
```
We already have an internal variable [`isRecordMode`](https://github.com/ndtp/android-testify/blob/2610898f25c8e42e5914858a6734bda6264b7ad5/Plugins/Gradle/src/main/kotlin/dev/testify/tasks/main/ScreenshotRecordTask.kt#L35C18-L35C30) being used by the `ScreenshotRecord` task which we can relocate to the `TestifySettings`. Then, we can configure that flag dynamically, without need to modify your code to record a baseline.

Reply:
Then this would work globally, though when setting via rule we can set it just for some tests, not for all. WDYT?

Reply:
Yes, that's a good point about being global. I think it comes down to what kind workflow the developer would have when writing tests.
When I made my comment, I was thinking about running this on a CI service -- that you'd want to set that flag via an environment variable that would control "record mode" globally so that you could update your screenshots using a CI service.

But, your question brings up a very valid flaw in my logic. What I think that we need is a convenient way to record a screenshot without needing to modify the code. I can't think of such a way right now, so I believe now that your current solution is acceptable and the best we can do for right now.

I think some of this could be addressed by updating both `screenshotTest` and `screenshotRecord` to work with the GMD gradle tasks so that they could configure Testify to work as expected. Also, there's some potential to update the IntelliJ Platform Plugin to support the GMD tasks too.

In any case, let's go with what you have now and we can always improve or enhance this later.

Reply:
Related to the above, can you describe your typical workflow when using GMD? As I've never used them myself, I'm curious how they fit in your development cycle.

Reply:
We use them to run screenshot tests on CI - it's easier to write few lines in `build.gradle`, than managing emulator. But when we need to update existing or create new baseline we do it locally

Reply:
I think I will add this parameter to `TestifyExtension` and will keep the way to enable record mode in the rule too

Reply:
Good idea